### PR TITLE
Neeraj ping provider

### PR DESCRIPTION
**Pull request by drekle**
These changes are from
https://github.com/neermitt/oauth2_proxy/
to include the ping provider.

### CODE REVIEW

**Comment by darkyat** [comment]
@josdotso, @dcwangmit01, @rtluckie - please approve.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+FROM golang:latest AS golang
+WORKDIR /go/src/github.com/bitly/oauth2_proxy
+RUN go get -u github.com/golang/dep/cmd/dep
+COPY . .
+#RUN dep ensure && \
```
Comment (ghost) [inline Dockerfile:5]: If this repo isn't compatible with `dep`, this needs to happen soon imo. I'll not block this PR with it though.  Please create an issue on this repo to track.

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
Approved, but @dcwangmit01 knows a lot of about oauth2-proxy, so will leave open for this review.

**Comment by darkyat** [comment]
@dcwangmit01 would be great to have this reviewed today. Thanks!

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+RUN go get ./...
+RUN CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o oauth2_proxy .
```
Comment (dcwangmit01) [inline Dockerfile:7]: SHOULD: Fix this.

The overloading of this Dockerfile is a very unusual pattern.  Usually the app image is not combined with the build image.  If we separated the two, the app image could get really small.

The app image could be really tiny, using alpine.  Using a larger image with dev tools installed increases the attack surface.

I would separate this out with a Makefile target that does a go build first, and then another Makefile target that does the docker build which copies the binary into an alpine image.  You need a makefile for a docker push anyways.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+FROM scratch
```
Comment (dcwangmit01) [inline Dockerfile:9]: Does the Dockerfile accept multiple FROM statements?

**Comment by dcwangmit01** [comment]
MUST: Explain:

1) Why the existing OIDC provider does not work for Ping Identity.
2) What the differences are between this new PingIdentity Provider and the existing OIDC Provider.

I'm not seeing any code differences.

From the looks of it, I think PingIdentity should work with the OIDC provider without this PR.

**Comment by dcwangmit01** [comment]
Also, everything that we do should be in a form that is upstreamable.  This PR should be upstreamed if it is necessary.

**Comment by ghost** [comment]
@drekle Please note that you should be able to create cisco-sso/oauth2_proxy PRs which pull from any public branch on forks not owned by you.  You should be able to PR using that method to avoid code re-publishing this way.

File: Dockerfile (hunk 3 above)
Comment (ghost) [reply Dockerfile:9]: His Dockerfile adopts the multi-stage feature added recently to Docker: https://docs.docker.com/develop/develop-images/multistage-build/#before-multi-stage-builds

File: Dockerfile (hunk 2 above)
Comment (dcwangmit01) [reply Dockerfile:7]: Sweet.  Actually, it is a good thing.  Thanks for the link @josdotso 

> His Dockerfile adopts the multi-stage feature added recently to Docker: https://docs.docker.com/develop/develop-images/multistage-build/#before-multi-stage-builds`

File: Dockerfile (hunk 3 above)
Comment (dcwangmit01) [reply Dockerfile:9]: Ack.

File: Dockerfile (hunk 2 above)
Comment (dcwangmit01) [reply Dockerfile:7]: This is good.

**Comment by drekle** [comment]
I believe Ezra also noted that this should be a part of the OIDC provider.  Not sure when/who should perform these changes.

Can you add Neeraj @neermitt

**Pull request closed without merging**